runtime/lua/vim/fcnotify.lua

```diff
@@ -0,0 +1,224 @@
+-- Create a file watcher, each watcher is identified by the name of the file that it
+-- watches.
+
+local uv = vim.loop
+local Watcher = {}
+Watcher.__index = Watcher
+local WatcherList = {}
+local check_handle = nil
+
+-- Checks if a buffer should have a watcher attached to it.
+local function valid_buf(bufnr)
+  if not vim.api.nvim_buf_is_valid(bufnr) then
+    return false
+  end
+
+  local buflisted = vim.api.nvim_buf_get_option(bufnr, 'buflisted')
+  local buftype = vim.api.nvim_buf_get_option(bufnr, 'buftype')
+
+  return buflisted or buftype == '' or buftype == 'acwrite'
+end
+
+-- Callback for the check handle, checks if there are pending notifications
+-- for any watcher, and handles them as per the value of the `fcnotify`
+-- option.
+local function check_notifications()
+  for _, watcher in pairs(WatcherList) do
+    if watcher.pending_notifs == true and valid_buf(watcher.bufnr) then
+      vim.api.nvim_command(string.format('checktime %d', watcher.bufnr))
+      watcher.pending_notifs = false
+    end
+  end
+end
+
+local function fs_event_start(bufnr)
+  WatcherList[bufnr].handle = uv.new_fs_event()
+  WatcherList[bufnr].handle:start(WatcherList[bufnr].fpath, {}, vim.schedule_wrap(function(...)
+    WatcherList[bufnr]:on_change(...)
+  end))
```

Comment:
It'd be nice if this just did something like `local watcher = WatcherList[bufnr]` and then you can use that `watcher` in this function rather than repeating it every time.

Reply:
(If that still works. The function currently looks that up every time it runs the function, so maybe the logic isn't right)